### PR TITLE
Fix #6334: githubRelease.releaseAssets: migrate from hardcoded paths to cross-project artifact dependencies

### LINKED ISSUE #6334: `detekt-generator-1.23.1-all.jar` is missing
## Expected Behavior
https://github.com/detekt/detekt/releases/tag/v1.23.1 should have `detekt-generator-1.23.1-all.jar` as 1.23.0.

## Observed Behavior
The `detekt-generator-1.23.1-all.jar` is missing.

## Steps to Reproduce
I've noticed that the file is missing entirely:
1. Clone the repo
2. `git checkout release/1.x`
3. `./gradlew build`
4. `cat detekt-generator/build/libs/detekt-generator-1.23.1-all.jar` <- this will fail

## Context
I'm pinging the rest of @detekt/maintainers as I couldn't find what caused the shadow plugin to fail here, but this could be a symptom of an underlying issue.

### PR DESCRIPTION
**Pull request by TWiStErRob**
Fixes #6334

Testing: `gradlew githubRelease -Pgithub.token=... --console=verbose`

<details><summary>release/1.x</summary>

```
> Task :githubRelease
:githubRelease [This task is a dry run. All API calls that would modify the repo are disabled. API calls that access the repo information are not disabled. Use this to show what actions would be executed.]
:githubRelease [CHECKING FOR PREVIOUS RELEASE]
:githubRelease [EXISTING RELEASE FOUND https://github.com/detekt/detekt/releases/tag/v1.23.1]
:githubRelease [DELETING RELEASE v1.23.1]
:githubRelease [CREATING NEW RELEASE
{
    tag_name               = v1.23.1
    target_commitish       = main
    name                   = v1.23.1
    generate_release_notes = false
    body                   = <omitted>
    draft                  = true
    prerelease             = true
}]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-cli\build\libs\detekt-cli-1.23.1-all.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-cli\build\distributions\detekt-cli-1.23.1.zip]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-formatting\build\libs\detekt-formatting-1.23.1.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-generator\build\libs\detekt-generator-1.23.1-all.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-rules-libraries\build\libs\detekt-rules-libraries-1.23.1.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-rules-ruleauthors\build\libs\detekt-rules-ruleauthors-1.23.1.jar]

BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed

```

</details>

<details><summary>This branch</summary>

<details><summary>All the tasks that are added because the artifacts are now being built.</summary>

```
> Task :detekt-psi-utils:processResources NO-SOURCE
> Task :detekt-metrics:processResources NO-SOURCE
> Task :detekt-parser:processResources NO-SOURCE
> Task :detekt-api:processResources NO-SOURCE
> Task :detekt-rules:processResources NO-SOURCE
> Task :detekt-rules:compileKotlin NO-SOURCE
> Task :detekt-utils:processResources NO-SOURCE
> Task :detekt-tooling:processResources NO-SOURCE
> Task :detekt-rules-exceptions:processResources UP-TO-DATE
> Task :detekt-cli:processResources UP-TO-DATE
> Task :detekt-rules-errorprone:processResources UP-TO-DATE
> Task :detekt-report-txt:processResources UP-TO-DATE
> Task :detekt-rules-coroutines:processResources UP-TO-DATE
> Task :detekt-report-sarif:processResources UP-TO-DATE
> Task :detekt-rules-documentation:processResources UP-TO-DATE
> Task :detekt-rules-complexity:processResources UP-TO-DATE
> Task :detekt-rules-empty:processResources UP-TO-DATE
> Task :detekt-report-xml:processResources UP-TO-DATE
> Task :detekt-report-md:processResources UP-TO-DATE
> Task :detekt-generator:processResources NO-SOURCE
> Task :detekt-rules-naming:processResources UP-TO-DATE
> Task :detekt-core:processResources UP-TO-DATE
> Task :detekt-rules-performance:processResources UP-TO-DATE
> Task :detekt-rules-style:processResources UP-TO-DATE
> Task :detekt-report-html:processResources UP-TO-DATE
> Task :detekt-rules-libraries:sourcesJar UP-TO-DATE
> Task :detekt-rules:compileJava NO-SOURCE
> Task :detekt-rules-ruleauthors:sourcesJar UP-TO-DATE
> Task :detekt-rules:classes UP-TO-DATE
> Task :detekt-utils:compileKotlin UP-TO-DATE
> Task :detekt-psi-utils:compileKotlin UP-TO-DATE
> Task :detekt-utils:compileJava NO-SOURCE
> Task :detekt-psi-utils:compileJava NO-SOURCE
> Task :detekt-rules:jar UP-TO-DATE
> Task :detekt-psi-utils:classes UP-TO-DATE
> Task :detekt-utils:classes UP-TO-DATE
> Task :detekt-utils:jar UP-TO-DATE
> Task :detekt-psi-utils:jar UP-TO-DATE
> Task :detekt-parser:compileKotlin UP-TO-DATE
> Task :detekt-api:compileKotlin UP-TO-DATE
> Task :detekt-parser:compileJava NO-SOURCE
> Task :detekt-api:compileJava NO-SOURCE
> Task :detekt-api:classes UP-TO-DATE
> Task :detekt-parser:classes UP-TO-DATE
> Task :detekt-api:jar UP-TO-DATE
> Task :detekt-parser:jar UP-TO-DATE
> Task :detekt-report-txt:compileKotlin UP-TO-DATE
> Task :detekt-report-xml:compileKotlin UP-TO-DATE
> Task :detekt-rules-ruleauthors:compileKotlin UP-TO-DATE
> Task :detekt-metrics:compileKotlin UP-TO-DATE
> Task :detekt-rules-coroutines:compileKotlin UP-TO-DATE
> Task :detekt-rules-documentation:compileKotlin UP-TO-DATE
> Task :detekt-rules-empty:compileKotlin UP-TO-DATE
> Task :detekt-tooling:compileKotlin UP-TO-DATE
> Task :detekt-rules-naming:compileKotlin UP-TO-DATE
> Task :detekt-rules-libraries:compileKotlin UP-TO-DATE
> Task :detekt-rules-performance:compileKotlin UP-TO-DATE
> Task :detekt-rules-exceptions:compileKotlin UP-TO-DATE
> Task :detekt-formatting:compileKotlin UP-TO-DATE
> Task :detekt-report-txt:compileJava NO-SOURCE
> Task :detekt-report-xml:compileJava NO-SOURCE
> Task :detekt-report-xml:classes UP-TO-DATE
> Task :detekt-rules-coroutines:compileJava NO-SOURCE
> Task :detekt-rules-exceptions:compileJava NO-SOURCE
> Task :detekt-rules-documentation:compileJava NO-SOURCE
> Task :detekt-formatting:compileJava NO-SOURCE
> Task :detekt-rules-performance:compileJava NO-SOURCE
> Task :detekt-rules-naming:compileJava NO-SOURCE
> Task :detekt-report-txt:classes UP-TO-DATE
> Task :detekt-rules-ruleauthors:compileJava NO-SOURCE
> Task :detekt-rules-coroutines:classes UP-TO-DATE
> Task :detekt-rules-documentation:classes UP-TO-DATE
> Task :detekt-tooling:compileJava NO-SOURCE
> Task :detekt-tooling:classes UP-TO-DATE
> Task :detekt-metrics:compileJava NO-SOURCE
> Task :detekt-metrics:classes UP-TO-DATE
> Task :detekt-rules-exceptions:classes UP-TO-DATE
> Task :detekt-rules-performance:classes UP-TO-DATE
> Task :detekt-rules-naming:classes UP-TO-DATE
> Task :detekt-rules-empty:compileJava NO-SOURCE
> Task :detekt-rules-libraries:compileJava NO-SOURCE
> Task :detekt-rules-documentation:jar UP-TO-DATE
> Task :detekt-rules-empty:classes UP-TO-DATE
> Task :detekt-rules-coroutines:jar UP-TO-DATE
> Task :detekt-rules-performance:jar UP-TO-DATE
> Task :detekt-report-xml:jar UP-TO-DATE
> Task :detekt-rules-exceptions:jar UP-TO-DATE
> Task :detekt-metrics:jar UP-TO-DATE
> Task :detekt-tooling:jar UP-TO-DATE
> Task :detekt-report-sarif:compileKotlin UP-TO-DATE
> Task :detekt-report-txt:jar UP-TO-DATE
> Task :detekt-cli:compileKotlin UP-TO-DATE
> Task :detekt-rules-errorprone:compileKotlin UP-TO-DATE
> Task :detekt-rules-naming:jar UP-TO-DATE
> Task :detekt-rules-empty:jar UP-TO-DATE
> Task :detekt-report-sarif:compileJava NO-SOURCE
> Task :detekt-cli:compileJava NO-SOURCE
> Task :detekt-cli:classes UP-TO-DATE
> Task :detekt-rules-errorprone:compileJava NO-SOURCE
> Task :detekt-report-sarif:classes UP-TO-DATE
> Task :detekt-report-html:compileKotlin UP-TO-DATE
> Task :detekt-rules-errorprone:classes UP-TO-DATE
> Task :detekt-cli:jar UP-TO-DATE
> Task :detekt-rules-style:compileKotlin UP-TO-DATE
> Task :detekt-rules-complexity:compileKotlin UP-TO-DATE
> Task :detekt-report-md:compileKotlin UP-TO-DATE
> Task :detekt-report-sarif:jar UP-TO-DATE
> Task :detekt-report-html:compileJava NO-SOURCE
> Task :detekt-report-html:classes UP-TO-DATE
> Task :detekt-report-md:compileJava NO-SOURCE
> Task :detekt-rules-complexity:compileJava NO-SOURCE
> Task :detekt-report-md:classes UP-TO-DATE
> Task :detekt-rules-style:compileJava NO-SOURCE
> Task :detekt-rules-style:classes UP-TO-DATE
> Task :detekt-rules-errorprone:jar UP-TO-DATE
> Task :detekt-report-html:jar UP-TO-DATE
> Task :detekt-generator:compileKotlin UP-TO-DATE
> Task :detekt-rules-complexity:classes UP-TO-DATE
> Task :detekt-report-md:jar UP-TO-DATE
> Task :detekt-rules-style:jar UP-TO-DATE
> Task :detekt-generator:compileJava NO-SOURCE
> Task :detekt-rules-complexity:jar UP-TO-DATE
> Task :detekt-generator:classes UP-TO-DATE
> Task :detekt-core:compileKotlin UP-TO-DATE
> Task :detekt-core:compileJava NO-SOURCE
> Task :detekt-core:classes UP-TO-DATE
> Task :detekt-core:jar UP-TO-DATE
> Task :detekt-generator:shadowJar UP-TO-DATE
> Task :detekt-cli:shadowJar UP-TO-DATE
> Task :detekt-generator:generateDocumentation UP-TO-DATE
> Task :detekt-cli:startShadowScripts UP-TO-DATE
> Task :detekt-rules-ruleauthors:processResources UP-TO-DATE
> Task :detekt-rules-libraries:processResources UP-TO-DATE
> Task :detekt-rules-libraries:classes UP-TO-DATE
> Task :detekt-cli:shadowDistZip UP-TO-DATE
> Task :detekt-rules-ruleauthors:classes UP-TO-DATE
> Task :detekt-formatting:processResources UP-TO-DATE
> Task :detekt-formatting:classes UP-TO-DATE
> Task :detekt-rules-libraries:jar UP-TO-DATE
> Task :detekt-rules-ruleauthors:jar UP-TO-DATE
> Task :detekt-formatting:jar UP-TO-DATE
```

</details>

```
> Task :githubRelease
:githubRelease [This task is a dry run. All API calls that would modify the repo are disabled. API calls that access the repo information are not disabled. Use this to show what actions would be executed.]
:githubRelease [CHECKING FOR PREVIOUS RELEASE]
:githubRelease [EXISTING RELEASE FOUND https://github.com/detekt/detekt/releases/tag/v1.23.1]
:githubRelease [DELETING RELEASE v1.23.1]
:githubRelease [CREATING NEW RELEASE
{
    tag_name               = v1.23.1
    target_commitish       = main
    name                   = v1.23.1
    generate_release_notes = false
    body                   = <omitted>
    draft                  = true
    prerelease             = true
}]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-cli\build\libs\detekt-cli-1.23.1-all.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-cli\build\distributions\detekt-cli-1.23.1.zip]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-generator\build\libs\detekt-generator-1.23.1-all.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-formatting\build\libs\detekt-formatting-1.23.1.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-rules-libraries\build\libs\detekt-rules-libraries-1.23.1.jar]
:githubRelease [UPLOADING P:\projects\contrib\github-detekt\detekt-rules-ruleauthors\build\libs\detekt-rules-ruleauthors-1.23.1.jar]

BUILD SUCCESSFUL in 1s
79 actionable tasks: 1 executed, 78 up-to-date
```

</details>

### CODE REVIEW

File: build-logic/src/main/kotlin/releasing.gradle.kts [hunk 1]
```diff
@@ -37,17 +42,31 @@ githubRelease {
...
+        isTransitive = false
+    }
+    releaseArtifacts(project(":detekt-rules-libraries")) {
+        targetConfiguration = Dependency.DEFAULT_CONFIGURATION
+        isTransitive = false
+    }
+    releaseArtifacts(project(":detekt-rules-ruleauthors")) {
+        targetConfiguration = Dependency.DEFAULT_CONFIGURATION
+        isTransitive = false
+    }
```
Comment (TWiStErRob) [inline build-logic/src/main/kotlin/releasing.gradle.kts:69]: `isTransitive = false` is weird, but that's what gives the wanted output.

File: build-logic/src/main/kotlin/releasing.gradle.kts [hunk 2]
```diff
@@ -37,17 +42,31 @@ githubRelease {
...
+    releaseAssets.setFrom(releaseArtifacts)
+}
+
+dependencies {
+    releaseArtifacts(project(":detekt-cli")) {
+        targetConfiguration = "shadow" // com.github.jengelman.gradle.plugins.shadow.ShadowBasePlugin.CONFIGURATION_NAME
+    }
+    releaseArtifacts(project(":detekt-cli")) {
+        targetConfiguration = "shadowDist"
+    }
```
Comment (TWiStErRob) [inline build-logic/src/main/kotlin/releasing.gradle.kts:54]: This was the only artifact that was not automatically created in some way already.

**Review by cortinico — APPROVED** [review]
Thanks for looking inot this @TWiStErRob

**Comment by TWiStErRob** [comment]
@cortinico is it targeting the right branch? After merge I guess we need to pick it to master too.

**Comment by cortinico** [comment]
> @cortinico is it targeting the right branch? After merge I guess we need to pick it to main too.

I'd say let's move this to `main` and I'll pick it on the release branch afterwards once it's green there 👍

**Comment by TWiStErRob** [comment]
Rebased using IDEA's magic wand and changed base branch.

**Review by cortinico — APPROVED** [review]

**Review by BraisGabin — APPROVED** [review]

**Comment by BraisGabin** [comment]
Something strange is going on here. @TWiStErRob could you rebase to force ci to run?

**Pull request merged**